File: python/meetup/meetup.py
import datetime
import calendar


WEEKDAYS = {
    "Monday": calendar.MONDAY,
    "Tuesday": calendar.TUESDAY,
    "Wednesday": calendar.WEDNESDAY,
    "Thursday": calendar.THURSDAY,
    "Friday": calendar.FRIDAY,
    "Saturday": calendar.SATURDAY,
    "Sunday": calendar.SUNDAY,
}

ORDS = {"first": 0, "second": 7, "third": 14, "fourth": 21, "fifth": 28, "last": None}


class MeetupDayException(Exception):
    def __init__(self, message):
        self.message = message



def meetup(year, month, week, day_of_week):
    _, num_days = calendar.monthrange(year, month)
    if week in list(ORDS.keys()):
        for day in range(1, 8):
            weekday = calendar.weekday(year, month, day)
            if weekday == WEEKDAYS[day_of_week]:
                if week == "last":
                    is_5th_the_last_week = day + ORDS["fifth"] <= num_days
                    return datetime.date(
                        year, 
                        month, 
                        day + ORDS["fifth" if is_5th_the_last_week else "fourth"]
                    )
                else:
                    if day + ORDS[week] <= num_days:
                        return datetime.date(year, month, day + ORDS[week])
    
    elif week == "teenth":
        for day in range(13, 20):
            weekday = calendar.weekday(year, month, day)
            if weekday == WEEKDAYS[day_of_week]:
                return datetime.date(year, month, day)
     
    raise MeetupDayException("That day does not exist.")

File: go/meetup/meetup.go
package meetup

import "time"

// Define the WeekSchedule type here.
type WeekSchedule int

const (
	First  WeekSchedule = 0
	Second WeekSchedule = 7
	Third  WeekSchedule = 14
	Fourth WeekSchedule = 21
	Fifth  WeekSchedule = 28
	Last   WeekSchedule = iota
	Teenth
)

var incDays = map[WeekSchedule]WeekSchedule{
	First:  First,
	Second: Second,
	Third:  Third,
	Fourth: Fourth,
	Fifth:  Fifth,
	Last:   Last,
}

var Weekdays = map[string]time.Weekday{
	"Monday":    time.Monday,
	"Tuesday":   time.Tuesday,
	"Wednesday": time.Wednesday,
	"Thursday":  time.Thursday,
	"Friday":    time.Friday,
	"Saturday":  time.Saturday,
	"Sunday":    time.Sunday,
}

func numDaysInMonth(m time.Month, y int) int {
	return time.Date(y, m+1, 0, 0, 0, 0, 0, time.UTC).Day()
}

func Day(week WeekSchedule, dayOfWeek time.Weekday, month time.Month, year int) int {
	if week == Teenth {
		for d := 13; d < 20; d++ {
			genDay := time.Date(year, month, d, 0, 0, 0, 0, time.UTC)
			if genDay.Weekday() == Weekdays[dayOfWeek.String()] {
				return genDay.Day()
			}
		}
	} else {
		if _, ok := incDays[week]; ok {
			for d := 1; d < 8; d++ {
				numDays := numDaysInMonth(month, year)
				genDay := time.Date(year, month, d, 0, 0, 0, 0, time.UTC)
				if genDay.Weekday() == Weekdays[dayOfWeek.String()] {
					if week == Last {
						is5thTheFinalWeek := d+int(Fifth) <= numDays
						if is5thTheFinalWeek {
							return time.Date(year, month, d+int(Fifth), 0, 0, 0, 0, time.UTC).Day()
						} else {
							return time.Date(year, month, d+int(Fourth), 0, 0, 0, 0, time.UTC).Day()
						}
					} else {
						addDaysRemainsTheSameMonth := d+int(incDays[week]) <= numDays
						if addDaysRemainsTheSameMonth {
							return time.Date(year, month, d+int(incDays[week]), 0, 0, 0, 0, time.UTC).Day()
						}
					}
				}
			}
		}
	}
	return 0
}
